# Movies Tonight Part 3: SQL DML

In [1]:
%load_ext sql
%sql sqlite:///Movies.db

'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

Remember!
- use this to get data from csv file- open terminal - sqlite3 Movies.db - follow "here" link
- .table will give list of tables- can check for duplicates!!
- PRAGMA will give metadata

In [2]:
%%sql

SELECT * FROM DATASET LIMIT 10;
PRAGMA table_info(DATASET);

 * sqlite:///Movies.db
Done.
Done.


[(0, 'TName', 'TEXT', 0, None, 0),
 (1, 'Location', 'TEXT', 0, None, 0),
 (2, 'Phone', 'TEXT', 0, None, 0),
 (3, 'MTitle', 'TEXT', 0, None, 0),
 (4, 'ShowTime', 'TEXT', 0, None, 0),
 (5, 'Rating', 'TEXT', 0, None, 0),
 (6, 'CCode', 'TEXT', 0, None, 0),
 (7, 'CName', 'TEXT', 0, None, 0)]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [21]:
%%sql

DELETE FROM MOVIES;
SELECT DISTINCT MTitle, Rating FROM DATASET
LIMIT 10;

 * sqlite:///Movies.db
23 rows affected.
Done.


[('Associate, The', 'PG-13'),
 ('Ghost & The Darkness, The', 'R'),
 ('Independence Day', 'PG-13'),
 ('D3: The Mighty Ducks', 'PG'),
 ('Dear God', ''),
 ('First Wives Club, The', 'PG-13'),
 ('High School High', 'PG-13'),
 ('Larger Than Life', 'PG'),
 ('Mirror Has Two Faces, The', 'PG-13'),
 ('Ransom', 'R')]

lets insert into the table!
REMEMBER: INSERT INTO table (part 2 titles)
          SELECT DISTINCT (columns from DATASET)
Shows- joining the other two tables to retrieve data for the shows table

In [13]:
%%sql

DELETE FROM Movies;
INSERT INTO Movies (Title, Rating)
    SELECT DISTINCT MTitle, Rating FROM DATASET;
SELECT * FROM MOVIES
LIMIT 10;

 * sqlite:///Movies.db
23 rows affected.
23 rows affected.
Done.


[(1, 'Associate, The', 'PG-13'),
 (2, 'Ghost & The Darkness, The', 'R'),
 (3, 'Independence Day', 'PG-13'),
 (4, 'D3: The Mighty Ducks', 'PG'),
 (5, 'Dear God', ''),
 (6, 'First Wives Club, The', 'PG-13'),
 (7, 'High School High', 'PG-13'),
 (8, 'Larger Than Life', 'PG'),
 (9, 'Mirror Has Two Faces, The', 'PG-13'),
 (10, 'Ransom', 'R')]

In [14]:
%%sql
DELETE FROM Theaters;
INSERT INTO Theaters (Name, Location, Phone)
    SELECT DISTINCT TName, Location, Phone FROM DATASET;
SELECT * FROM Theaters;

 * sqlite:///Movies.db
0 rows affected.
9 rows affected.
Done.


[(1, 'Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899'),
 (2, 'Cinema Star The Ultraplex 14', 'Mission Grove, Riverside', '(909) 342-2256'),
 (3, 'General Cinema Rancho 6', ' I-215 At Mt. Vernon S. At I-10, San Bernardino', '(714) 370-2085'),
 (4, 'Pacific Inland Center', ' Inland Center Mall, San Bernardino', '(714) 381-1611'),
 (5, 'SOCAL Canyon Crest Cinema', 'Central Avenue South Of 60 Freeway Near Ucr, Riverside', '(909) 682-6900'),
 (6, 'SOCAL Canyon Springs Cinema', 'East Of I-215 On 60 Freeway At Day Street Canyon, Moreno Valley', '(909) 782-0800'),
 (7, 'SOCAL Marketplace Cinema', 'University/mission Inn Exits East Of 91 Freeway On, Riverside', '(909) 682-4040'),
 (8, 'United Artists Riverside (Galleria) Tyler Mall', 'Riverside Fwy Tyler, Riverside', '(714) 689-802'),
 (9, 'United Artists Riverside Park Sierra', '3600 Park Sierra Dr., Riverside', '(909) 359-6995')]

In [18]:
%%sql
DELETE FROM Artists;
INSERT INTO Artists (Name)
    SELECT DISTINCT CName FROM DATASET;
SELECT * FROM Artists
LIMIT 10;

 * sqlite:///Movies.db
153 rows affected.
153 rows affected.
Done.


[(1, 'Austin Pendleton', None, None),
 (2, 'Bebe Neuwirth', None, None),
 (3, 'Dianne Wiest', None, None),
 (4, 'Eli Wallach', None, None),
 (5, 'Kenny Kerr', None, None),
 (6, 'Lainie Kazan', None, None),
 (7, 'Tim Daly', None, None),
 (8, 'Whoopi Goldberg', None, None),
 (9, 'Donald Petrie', None, None),
 (10, 'Bernard Hill', None, None)]

In [19]:
%%sql
DELETE FROM Shows;
INSERT INTO Shows (ShowTime, MID, TID)
    SELECT DISTINCT DATASET.ShowTime, MOVIES.MID, THEATERS.TID
    FROM DATASET
        JOIN MOVIES ON (MOVIES.Title = DATASET.MTitle)
        JOIN THEATERS ON (THEATERS.Name = DATASET.TName);
SELECT ShowTime, Title, Name FROM Shows JOIN Movies USING (MID) JOIN Theaters USING (TID)
LIMIT 10;

 * sqlite:///Movies.db
131 rows affected.
131 rows affected.
Done.


[('4:20 PM', 'Associate, The', 'Akarakian Theatres Moreno 4 Cinemas'),
 ('7:20 PM', 'Associate, The', 'Akarakian Theatres Moreno 4 Cinemas'),
 ('9:40 PM', 'Associate, The', 'Akarakian Theatres Moreno 4 Cinemas'),
 ('5:10 PM', 'Ghost & The Darkness, The', 'Akarakian Theatres Moreno 4 Cinemas'),
 ('9:00 PM', 'Ghost & The Darkness, The', 'Akarakian Theatres Moreno 4 Cinemas'),
 ('9:05 PM', 'Ghost & The Darkness, The', 'Akarakian Theatres Moreno 4 Cinemas'),
 ('7:00 PM', 'Independence Day', 'Akarakian Theatres Moreno 4 Cinemas'),
 ('4:40 PM', 'Associate, The', 'Cinema Star The Ultraplex 14'),
 ('7:30 PM', 'Associate, The', 'Cinema Star The Ultraplex 14'),
 ('2:30 PM', 'D3: The Mighty Ducks', 'Cinema Star The Ultraplex 14')]

In [20]:
%%sql
DELETE FROM Credit;
INSERT INTO Credit (CCode, AID, MID)
    SELECT DISTINCT DATASET.CCode, MOVIES.MID, ARTISTS.AID
    FROM DATASET
        JOIN ARTISTS ON (ARTISTS.NAME = DATASET.CName)
        JOIN MOVIES ON (MOVIES.Title = DATASET.MTitle);
SELECT * FROM CREDIT 
LIMIT 10;

 * sqlite:///Movies.db
162 rows affected.
162 rows affected.
Done.


[(1, 1, 1, 'A'),
 (2, 1, 2, 'A'),
 (3, 1, 3, 'A'),
 (4, 1, 4, 'A'),
 (5, 1, 5, 'A'),
 (6, 1, 6, 'A'),
 (7, 1, 7, 'A'),
 (8, 1, 8, 'A'),
 (9, 1, 9, 'D'),
 (10, 2, 10, 'A')]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

In [ ]:
# checked in cells above

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [29]:
%%sql
/* Found using https://chartio.com/learn/databases/how-to-find-duplicate-values-in-a-sql-table/ */
    
SELECT MTitle, TName, Showtime, CCode, CName, COUNT(*) occurrences
FROM DATASET
GROUP BY MTitle, TName, Showtime, CCode, CName
HAVING 
    COUNT(*) > 1;

 * sqlite:///Movies.db
Done.


[('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Greg Kinnear', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Hector Elizondo', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Jon Seda', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Laurie Metcalf', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Maria Pitillo', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Roscoe Lee Browne', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Tim Conway', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'D', 'Garry Marshall', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Jon Lovitz', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Mekhi Phifer', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Tia Carrera', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'and Louise Fletcher', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'D', 'Hart Bochner', 2)]

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

In [30]:
%%sql
DROP TABLE IF EXISTS DATASET

 * sqlite:///Movies.db
Done.


[]

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".